accepted/wasmtime-baseline-compilation.md

```diff
@@ -0,0 +1,259 @@
+within Cranelift. We can continue to use ISLE for these, but they will need to
+be moved to the separate crate.
+
+
+As a result of this initial layering, one will be able to build a `MachInst` as
+a Rust data structure and emit it manually, for example:
+
+```rust
+let add = cranelift_asm::x64::AluRmiR { op: AluRmiR::Add, … };
+let mut buf = MachBuffer::new();
+add.emit(&mut buf, …);
+```
+
+However this is still quite cumbersome. As a next step, we will develop an API
+over this that provides for procedural generation of instructions: i.e., one
+method call for each instruction. Something like:
+
+```rust
+let mut masm = cranelift_asm::x64::MacroAssembler::new(); 
+masm.add(rd, rm);
+masm.store(rd, MemArg::base_offset(ra, 64));
+``` 
+This would allow for
+fairly natural single-pass code emission. In essence, this is an implementation
+of the [MacroAssembler
+idea](https://searchfox.org/mozilla-central/rev/fa71140041c5401b80a11f099cc0cd0653295e2c/js/src/jit/MacroAssembler.h)
+from SpiderMonkey. Each architecture will have an implementation of the
+MacroAssembler API; perhaps there can be a trait that abstracts commonalities,
+but enough will be different (e.g., instruction set quirks beyond the usual
+“add/sub/and/or/not” suspects, x64 two-operand form vs aarch64 three-operand
+form, and more) that we expect there to be different `MacroAssembler` types for
+each ISA. This in turn implies different lowering code that invokes the
+`MacroAssembler` per ISA in the baseline compiler. The lowering code can perhaps
+share many helpers that are monomorphized on the “common ISA core” trait.
+
+In the above examples, we bypass the register-allocation support, i.e. the
+ability to hold virtual register operands rather than real registers, in the
+`MachInst`s. This is supported today by passing through `RealReg`s (“real
+registers”) instead. In the baseline compiler we expect register allocation to
+occur before invoking the `MacroAssembler`; i.e., when generating the
+instructions we already know which register we are using for each operand. Doing
+otherwise (emitting with vregs first and editing later) requires actually
+buffering the MachInst structs in memory, which we do not wish to do.
+
+We don’t expect to make any changes to Cranelift itself beyond the layering
+refactor to borrow its `MachInst` and `MachBuffer` implementations. In
+particular we don’t expect to use the `MacroAssembler` wrapper in Cranelift, at
+least at first, because it will be built around constructing and emitting
+instructions to machine code right away, without buffering (as in Cranelift’s
+VCode). It’s possible in the future that we may find other ways to make
+`MacroAssembler` generic and leverage it in Cranelift too, but that is beyond
+the scope of this RFC.
+
+### Register Allocation
+
+We plan to implement register allocation in a linear[^2] fashion.
+
+[^2]: Known as [Linear Scan Register
+Allocation](http://web.cs.ucla.edu/~palsberg/course/cs132/linearscan.pdf)
```

Comment:
There are many variants of linear scan register allocation, and I'm afraid that referring to a textbook implementation may be confusing: here it's really allocating registers on the fly, and spilling aggressively whenever none is available, rather than doing any kind of textbook analysis to figure out liveness intervals and all of that.

Reply:
Agree, this reference should probably be removed @saulecabrera ; what we describe here isn't true linear-scan but something more like "single-pass register allocation" (proper linear scan requires lookahead to know which register is used furthest in the future, to choose which to evict).

Reply:
Agreed, I've removed the reference. For my own knowledge: are there any papers that describe each of the linear scan variants? Or are variations purely a result of implementation differences?

Reply:
I'm not aware of any that describe the SpiderMonkey-style single pass allocator, but I haven't looked too hard either; at the level we care about it tends to be more instructive to look at other implementations and academics tend not to study the fiddly details that matter here. But if I find one I'll let you know (and vice-versa) :-)